### PR TITLE
Add mm.lic, A general moon mage utility script

### PR DESCRIPTION
**Pull request by jandersson**
The initial commit contains the locate function, which locates a player
and displays the room number if the locate was successful.

### CODE REVIEW

File: mm.lic [hunk 1]
```diff
@@ -0,0 +1,61 @@
+=begin
+  Documentation:
+=end
+custom_require.call(%w[common common-arcana])
```
Comment (jandersson) [inline mm.lic]: Going to remove common, i just realized i added this but never used it

File: mm.lic [hunk 2]
```diff
@@ -0,0 +1,61 @@
+=begin
+  Documentation:
```
Comment (jandersson) [inline mm.lic]: Ill make the epedia doc and update this line with the link

**Comment by jandersson** [comment]
Just a quick usage example
```
P>;mm locate wabbajack                                                                      
--- Lich: mm active.
[mm]>release mana                                                                           
You aren't harnessing any mana.                                                             
[mm]>prepare cv 1                                                                           
That won't affect your current attunement very much.                                        
You spread your hands apart then slowly bring them together, fingers interlocked.           
[mm]>cast                                                                                   
You gesture.                                                                                
You feel more aware of your environment.                                                      
[mm]>release mana                                                                             
You aren't harnessing any mana.                                                               
[mm]>prepare locate 40                                                                        
That won't affect your current attunement very much.                                          
You spread your hands apart then slowly bring them together, fingers interlocked.             
P>                                                                                            
You feel fully attuned to the mana streams again.                                             
P>                                                                                            
 * Aarhhrsksha returns home from a hard day of adventuring.                                   
[mm]>cast wabbajack                                                                           
You gesture.                                                                                  
There doesn't seem to be anyone by that name under the skies.                                 
P>                                                                                            
[mm: Wabbajack is not in-game]                                                                
```
```
--- Lich: mm active.                                                                      
[mm]>release mana                                                                         
You aren't harnessing any mana.                                                           
[mm]>prepare cv 1                                                                         
That won't affect your current attunement very much.                                      
You spread your hands apart then slowly bring them together, fingers interlocked.         
[mm]>cast                                                                                 
You gesture.                                                                              
You feel more aware of your environment.                                                  
[mm]>release mana                                                                         
You aren't harnessing any mana.                                                           
[mm]>prepare locate 40                                                                    
That won't affect your current attunement very much.                                      
You spread your hands apart then slowly bring them together, fingers interlocked.         
P>                                                                                        
 * Shadow's Touch Mysterys Al'thor joins the adventure.                                   
P>                                                                                        
You feel fully attuned to the mana streams again.                                         
P>                                                                                        
 * Dyago returns home from a hard day of adventuring.                                     
[mm]>cast wabbajack                                                                       
You gesture.                                                                              
[The Crossing, Sicle Grove Lane]                                                          
A low, shed-like building stands beneath the trees along the road here.  It has a gabled r
  and tusks decorate the outside of the shed.                                             
Also here: Wabbajack.                                                                     
Obvious paths: east, west.                                                                
Room Number: 2603                                                                         
P>                                                                                        
[mm: Wabbajack appears to be in room #820]                                                
--- Lich: mm has exited.                                                                  
```

**Comment by jandersson** [comment]
Added rift functionality. I dont plan on making any more updates for the moment, unless its to correct a mistake.

File: mm.lic [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  def initialize
+    args = get_args
+    @settings = get_settings
+
+    if args.locate
+      locate?(args.person)
+    elsif args.rift
+      rift(args.person)
+    end
+    Flags.delete('locate-fail')
```
Comment (rpherbig) [inline mm.lic]: Should this deletion be done in `locate?`? I.e. it feels like delete should be in the same 'scope' as the add. Watch out for early returns though.

File: mm.lic [hunk 4]
```diff
@@ -0,0 +1,94 @@
+=begin
+  Documentation: https://elanthipedia.play.net/Lich_script_repository#mm
```
Comment (rpherbig) [inline mm.lic]: I'm not sure how I feel about using abbreviations to this degree in a script name... what do you think?

File: mm.lic [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  end
+
+  def rift(person)
+    person = person.capitalize
+    check_moonwatch
+    unless UserVars.moons.any? { |moon_name, moon_data| UserVars.moons['visible'].include?(moon_name) && moon_data['timer'] >= 0 }
+      echo 'No moons available to cast Rift'
+      return
+    end
+    data = if @settings.waggle_sets['rs']
```
Comment (rpherbig) [inline mm.lic]: I'm not sure how I feel about using abbreviations to this degree in a waggle set name... what do you think?

File: mm.lic [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+             @settings.waggle_sets['rs'].values.first
+           else
+             echo "Using discern values for RS. Make a waggle_set called 'rs' to avoid this in the future"
```
Comment (rpherbig) [inline mm.lic]: Do you think users would expect a waggle_set to be used by a script other than `buff.lic`?

File: mm.lic [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+             echo "Using discern values for locate. Make a waggle_set called 'locate' to avoid this in the future"
+             check_discern({ 'abbrev' => 'locate' }, @settings)
+           end
+    data['cast'] = "cast #{person}"
+    cast_spell(data, @settings)
+    pause 0.25
+    if Flags['locate-fail']
+      case Flags['locate-fail'].first
+      when /Your vision is not clear/
+        return locate?(person)
```
Comment (rpherbig) [inline mm.lic]: I'm not too familiar with MMs. Is it possible for this to loop forever, or is it very likely the locate will work on the next try?

**Comment by skyleronken** [comment]
Any plans to add teleport and moongate?

**Comment by skyleronken** [comment]
Also, might sound silly, but can we have the locate? method return the results as well? Im looking to use the functionality in a moonie bot script.

File: mm.lic (hunk 3 above)
Comment (jandersson) [reply mm.lic]: The early returns are why i moved it out of scope. I gave preference to only doing the delete on one line. In keeping with that preference, I could make `create_flags` `destroy_flags` methods which takes care of all the scripts flags.

File: mm.lic (hunk 4 above)
Comment (jandersson) [reply mm.lic]: I dont have a preference for the name. I went with something short like a command line tool. I could rename this to `;moonmage`.

File: mm.lic (hunk 5 above)
Comment (jandersson) [reply mm.lic]: Im following a naming convention for single spell sets where the name of the set is the spell abbreviation. This is a personal preference for convenience and lets all spells be cast via `;buff <abbrev>` for convenience. I am happy to use some other convention.

File: mm.lic (hunk 6 above)
Comment (jandersson) [reply mm.lic]: I wont speak for all users but there is precedent for it. Pick, astrology, zombie are three set names I can think of

File: mm.lic (hunk 7 above)
Comment (jandersson) [reply mm.lic]: It will loop forever if they fail to cast clear vision in 5 seconds (as `cast_spell` does)

Comment (jandersson) [reply mm.lic]: Ill have the script exit rather than retry.

**Comment by jandersson** [comment]
Yes teleport and moongate would be nice additions. The next step is to add some logic for tracking the state of your moonbeams as well as refreshing and creating them in specific rooms.

**Comment by jandersson** [comment]
I can have locate return the room number instead of `true` and `false` on failure. Let me know if this violates its 'predicateness', but i think all integers are truthy.

File: mm.lic (hunk 3 above)
Comment (rpherbig) [reply mm.lic]: That makes sense, I'm fine with whatever you think is best.

File: mm.lic (hunk 6 above)
Comment (rpherbig) [reply mm.lic]: The difference I'm seeing is that `pick` calls `buff` and just passes in the name of the waggle set. However, this uses the waggle set data directly. You're right that `astrology` does the same as `mm` though. I'm torn on this, so I'm willing to leave it alone and see what patterns emerge...

**Comment by jandersson** [comment]
> Which room number is correct?

The room number reported by the script. Roomnumbers reports your current room number when locating.

**Comment by rpherbig** [comment]
@jandersson looks good to me, let me know when you're ready for this to be merged

**Comment by jandersson** [comment]
Im happy with this iteration

**Pull request merged**